driver: flash_mspi_nor: Allow specific read/write IO modes and frequencies

utsavm9 opened the pull request:
Consider these MSPI flash parts: [Gigadevice GD25LE64E](https://www.mouser.com/datasheet/2/870/gd25le64e_rev1_3_20220426-2581248.pdf?srsltid=AfmBOorNVcnYnluZdrfwc0T8omPY6eENgh54MhcyPPMGnzULMow4bohG), [Macronix MX25U6432F](https://www.mxic.com.tw/Lists/Datasheet/Attachments/9025/MX25U6432F,%201.8V,%2064Mb,%20v1.1.pdf), [Winbond W25Q64JW](https://www.mouser.com/datasheet/2/949/W25Q64JW_RevB_11042019-1760399.pdf?srsltid=AfmBOoqU8en17EXwRIWMS5knHo8qgqum1mYZ11-1KLpk4Y5CnxYXuwyQ). If only D0 and D1 lines are connected to these flash parts, then Dual IO 1-1-2 `DREAD 0x3B` and Single IO `PP 0x02` are the best read and write commands available to use. Notice that neither Dual IO `PP_1_1_2 0xA2` command is supported in any of these flash parts, and neither these flash parts support Dual Peripheral Interface (like QPI, OPI), so all general commands need to be in Single IO.

The driver currently provides no way to use Dual IO Read and Single IO for the rest of the commands currently, and would erroneously use Single IO PP command in Dual IO mode. This PR fixes and adds support for that.

## Target Board
nRF54H20 MSPI and flash driver to support [Gigadevice GD25LE64E](https://www.mouser.com/datasheet/2/870/gd25le64e_rev1_3_20220426-2581248.pdf?srsltid=AfmBOorNVcnYnluZdrfwc0T8omPY6eENgh54MhcyPPMGnzULMow4bohG), [Macronix MX25U6432F](https://www.mxic.com.tw/Lists/Datasheet/Attachments/9025/MX25U6432F,%201.8V,%2064Mb,%20v1.1.pdf) and [Winbond W25Q64JW](https://www.mouser.com/datasheet/2/949/W25Q64JW_RevB_11042019-1760399.pdf?srsltid=AfmBOoqU8en17EXwRIWMS5knHo8qgqum1mYZ11-1KLpk4Y5CnxYXuwyQ).

## Modifications
*  JEDEC standard specifies different commands for read/writes in different MSPI IO modes. Driver should let device tree specify which IO mode and frequency to use for read and writes, defaulting to the overall IO mode and frequency when not specified.
* With different IO modes using different number of lines, one `mspi-max-frequency` is not the highest possible frequency to use for all modes. Allow read/write frequency to be granularly specified.

## Testing
The following modes for flash driver worked in local testing:
* Single IO for everything
```
mspi-io-mode = "MSPI_IO_MODE_SINGLE";
mspi-max-frequency = <DT_FREQ_M(66)>;
```
* Single IO general and write, Dual IO reads
```
mspi-io-mode = "MSPI_IO_MODE_SINGLE";
mspi-max-frequency = <DT_FREQ_M(66)>;
read-command = <0x3B>; // DREAD Dual IO 1-1-2
read-io-mode = "MSPI_IO_MODE_DUAL_1_1_2";
read-frequency = <DT_FREQ_M(40)>;
```
* Quad IO for writes and Dual IO for reads because why not.
```
mspi-max-frequency = <DT_FREQ_M(60)>;
mspi-io-mode = "MSPI_IO_MODE_SINGLE";
read-frequency = <DT_FREQ_M(30)>;
read-io-mode = "MSPI_IO_MODE_DUAL_1_1_2";
read-command = <0x3B>;
write-frequency = <DT_FREQ_M(40)>;
write-io-mode = "MSPI_IO_MODE_QUAD_1_4_4";
write-command = <0x38>;
```

swift-tk reviewed (changes requested):
You are adding write/read frequencies but not write/read io_modes to struct mspi_dev_cfg, why?
Anyway, I don't think this is the correct approach.
Please also include the specific devices in the description you are targeting so we can check the datasheets.

I do understand there might be a need for some flash devices that have different maximum frequency for write and read operation and even different io mode. But changing the mspi scope seems wrong, so maybe just add bindings to the flash scope(jedec,mspi-nor).

utsavm9 commented:
Re-did the PR after previous comments, might have oversimplied the IO mode selection logic while missing some flash mode transition, so requesting careful review. Works for all configs I can apply with my hardware setup. Commits are self-contained to help with review.

utsavm9 commented on drivers/flash/flash_mspi_nor.c:
```diff
@@ -997,21 +989,24 @@ static int flash_chip_init(const struct device *dev)
-
-	dev_data->in_target_io_mode = false;
+	dev_data->last_applied_cfg = &dev_data->mspi_control_cfg;
+	dev_data->mspi_control_cfg.freq = target_freq;
```
* Even in the original intention, we switch out of 50MHz by the time `switch_to_target_io()` was run for control commands too. Renamed `mspi_nor_init_cfg` to `mspi_control_cfg` because I want to have a MSPI dev config for all control (not read & write) commands like read status, erase etc. and not for use just during init. Perhaps `mspi_base_cfg` is a better name?
* Moved to data from config because the frequency is changed during runtime and cannot be const anymore. Frequency needs to be <= 50MHz only during initial commands like Read ID because I believe JEDEC compliance mandates supporting upto 50MHz. After init, we should use device tree specified frequency.

utsavm9 replied:
I'll revert it. Though, the code you highlighted on mainline would only pick Single IO out of `mspi_nor_init_cfg` to be applied with `mspi_dev_config` but this PR picks both IO mode and frequency out of the chosen configs, so that will result in init freq being applied which was not the case before.

swift-tk replied:
I meant we should keep `mspi_control_cfg` freq <= 50M so that when sending cmds we won't be meeting timing issue at target freq > 100MHz. This was not the case before as I missed that the `mspi_dev_config` only changes IO.

swift-tk commented on drivers/flash/flash_mspi_nor.c:
```diff
@@ -77,62 +84,44 @@ static uint16_t get_extended_command(const struct device *dev,
+			cfg = dev_data->write_cfg;
+		} else {
+			/* For all other commands, use control command config */
+			cfg = &dev_data->mspi_control_cfg;
```
I've lost track, does this change the driver behavior?  I think it does...
@anangl

utsavm9 replied:
It should not!
* All this is wrapped within a `!dev_config->multi_io_cmd` so drivers set with QPI or OPI as the IO mode skip this config selection logic anyways like mainline. They stay in QPI/OPI forever after initialization.
* Mainline drivers currently do not override read and write IO mode, so both read and write configs point to `mspi_nor_cfg` and that is applied just like on mainline for read and writes.
* For all control commands, just select Single IO mode always because codepath is not taken for QPI/OPI anyways. I believe the earlier logic accomplished the same, but instead of always selecting Single IO mode, it was trying to detect when Octal/Quad/Dual IO 1-1-x or 1-x-x specifically would not work.

A change in behavior could be that `mspi_dev_config` is called more times than before, but the communication with flash should stay the same for a device-tree config as far as I have tried.

swift-tk commented on drivers/flash/flash_mspi_nor.c:
```diff
@@ -77,62 +84,44 @@ static uint16_t get_extended_command(const struct device *dev,
+	if (cfg && cfg != dev_data->last_applied_cfg) {
 		rc = mspi_dev_config(dev_config->bus, &dev_config->mspi_id,
-				     MSPI_DEVICE_CONFIG_IO_MODE, cfg);
+			MSPI_DEVICE_CONFIG_IO_MODE | MSPI_DEVICE_CONFIG_FREQUENCY, cfg);
```
okay, I clearly miss it from before. Configuring the freqeuncy here is reasonable and aligned with the original intent.

anangl commented on drivers/flash/flash_mspi_nor.c:
```diff
@@ -1139,32 +1139,30 @@ static int flash_chip_init(const struct device *dev)
 			return rc;
 		}
-		dev_data->last_applied_cfg = NULL;
+		dev_data->last_applied_cfg = mspi_cfg;
```
I don't think this is needed. Neither IO mode nor frequency is reconfigured for XIP (see `XIP_DEV_CFG_MASK`), so all changes related to XIP can be dropped.

swift-tk replied:
I think it is useful for tracking the last applied cfg, especially during intermediate states of the device (Not in target IO mode(s)). Good to keep this to complete the logic (successful config leads to saving it).

utsavm9 committed:
drivers: flash_mspi_nor: Simplify perform_xfer function

Remove unnecessary function argument, makes later commits also simpler.

Signed-off-by: Utsav Munendra <utsavm@meta.com>

utsavm9 committed:
drivers: flash_mspi_nor: Track last applied MSPI config

Instead of just tracking in a bool whether the MSPI device is in
Standard MSPI vs. QPI/OPI config, track the entire MSPI config which was
last applied. This makes it easier later to track more than two configs
to apply based on the next command to transceive.

Signed-off-by: Utsav Munendra <utsavm@meta.com>

utsavm9 commented on drivers/flash/flash_mspi_nor.c:
```diff
@@ -1261,10 +1296,12 @@ static DEVICE_API(flash, drv_api) = {
 {									\
 	.ce_num = DT_INST_PROP_OR(inst, mspi_hardware_ce_num, 0),	\
-	.freq = MIN(DT_INST_PROP(inst, mspi_max_frequency), MHZ(50)),	\
+	.freq = FLASH_MSPI_MAX_FREQ(inst),				\
```
50MHz is not used for all register access on mainline right now. See your comment above where you saw that on mainline, only `MSPI_DEVICE_CONFIG_IO_MODE` is being applied in `perform_xfer()` on mainline from the config, causing register access to switch to `mspi-max-frequency` after initialization when `switch_to_target_io_mode()` is run.

This PR preserves that behavior of mainline. Whether or not to use 50MHz for all register access should be dealt with in a separate issue/PR as that is a behavior change. The cases you mentioned are valid, but I do not see how mainline supports that right now anyways.

anangl commented on drivers/flash/flash_mspi_nor.h:
```diff
@@ -115,6 +115,7 @@ struct flash_mspi_nor_data {
 	struct flash_mspi_nor_switch_info switch_info;
 	bool in_target_io_mode;
 	const struct mspi_dev_cfg *last_applied_cfg;
+	bool chip_detected;
```
I think this name is misleading. The driver does not perform a chip detection, it just configures the chip according to what is specified in dts. I'd rather use `chip_initialized` or `chip_configured`.

swift-tk commented on drivers/flash/flash_mspi_nor.c:
```diff
@@ -1158,6 +1164,7 @@ static int flash_chip_init(const struct device *dev)
 				rc);
 			return rc;
 		}
+		dev_data->last_applied_cfg = NULL;
```
There should have been a "different" flash_read API implementation that reads memory mapped bus address directly instead of going through mspi_transceive. Thus the XIP related configuration need to be configured during init if no mode switching needed or before read if there is mode switching.

utsavm9 replied:
That confirms my suspicion that no real XIP is supported by this driver right now, all flash_read() will be done by CPU anyways. This line should be dealt with whenever that is implemented, keeping it here for now to ensure CPU-based flash reads continues to work with mode switching.

utsavm9 committed:
drivers: flash_mspi_nor: Remove bool tracking of target IO mode

Remove the Boolean tracking of MSPI IO mode as we can now rely on
tracking the entire dev config applied to the MSPI device, multiple
of which will exist in later commits.

Signed-off-by: Utsav Munendra <utsavm@meta.com>

utsavm9 committed:
drivers: flash_mspi_nor: Flash control commands to use their own configs

Also in preparation for allowing control command frequency to be
different from the read/write frequency and initialization frequency.

Signed-off-by: Utsav Munendra <utsavm@meta.com>

utsavm9 committed:
driver: flash_mspi_nor: Allow specific read/write IO modes and frequencies

The driver currently provides no way to use Dual IO Read and Single IO
for the rest of the commands currently, and would erroneously use
Single IO PP command in Dual IO mode. This PR fixes and adds
support for that.

Signed-off-by: Utsav Munendra <utsavm@meta.com>

utsavm9 committed:
doc: release: Note changes to flash_mspi_nor

Note that read, write and control commands can be configured
separately via devicetree now.

Signed-off-by: Utsav Munendra <utsavm@meta.com>

anangl commented on doc/releases/release-notes-4.3.rst:
```diff
@@ -175,49 +175,15 @@ New APIs and options
+* Flash
+
+  * :dtcompatible:`jedec,mspi-nor` now allows configuration of read, write and
+   control commands separately via devicetree.
```
```suggestion
  * :dtcompatible:`jedec,mspi-nor` now allows configuration of read, write and
    control commands separately via devicetree.
```
I think this should fix the documentation build failure:
```
[1/2] Running Sphinx HTML build...
/__w/zephyr/zephyr/doc/_build/src/releases/release-notes-4.3.rst:181: WARNING: Bullet list ends without a blank line; unexpected unindent. [docutils]
FAILED: CMakeFiles/html /__w/zephyr/zephyr/doc/_build/CMakeFiles/html
```

jhedberg commented:
@utsavm9 is this for 4.3 or 4.4? Please set the milestone accordingly. If for 4.3, this also needs a matching bug report.
